mmcv/ops/csrc/pytorch/cuda/rans.cu

```diff
@@ -0,0 +1,265 @@
+      while ((raw_val >> (n_bypass * bypass_precision)) != 0) {
+        ++n_bypass;
+      }
+
+      /* Encode number of bypasses */
+      int32_t val = n_bypass;
+      while (val >= max_bypass_val) {
+        _syms.push_back({max_bypass_val, max_bypass_val + 1, true});
+        val -= max_bypass_val;
+      }
+      _syms.push_back(
+          {static_cast<uint16_t>(val), static_cast<uint16_t>(val + 1), true});
+
+      /* Encode raw value */
+      for (int32_t j = 0; j < n_bypass; ++j) {
+        const int32_t val =
+            (raw_val >> (j * bypass_precision)) & max_bypass_val;
+        _syms.push_back(
+            {static_cast<uint16_t>(val), static_cast<uint16_t>(val + 1), true});
+      }
+
+      while (!_syms.empty()) {
+        const RansSymbol sym = _syms.back();
+        Rans64EncPutBits(&rans, &ptr, sym.start, bypass_precision);
+        _syms.pop_back();
+      }
+    }
+    Rans64EncPut(&rans, &ptr, symbols[i].start, symbols[i].range, precision);
+  }
+
+  Rans64EncFlush(&rans, &ptr);
+
+  nbyte = std::distance(ptr, output.data() + output.size()) * sizeof(uint32_t);
+}
+
+std::string RansEncodeWithIndexesCUDAKernelLauncher(
+    const Tensor symbols, const Tensor indexes, const Tensor cdfs,
+    const Tensor cdfs_sizes, const Tensor offsets, int num_threads) {
+  // param symbols: (N, )
+  // param indexes: (N, )
+  // param cdfs: (M, max_cdf_bin_size)
+  // param cdfs_sizes: (M, )
+  // param offsets: (M, )
+
+  // check input
+  check_rans_encode_input(symbols, indexes, cdfs, cdfs_sizes, offsets,
+                          num_threads);
+
+  // set device
+  at::cuda::CUDAGuard device_guard(symbols.device());
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+
+  // count number of symbols
+  const int num_symbols = symbols.size(0);
+  const int num_cdfs = cdfs.size(0);
+  const int max_cdf_bin_size = cdfs.size(1);
+
+  // set kernel params
+  int cuda_threads = min(num_symbols, THREADS_PER_BLOCK);
+  dim3 blocks((num_symbols + cuda_threads - 1) / cuda_threads);
+  dim3 threads(cuda_threads);
+
+  // malloc device memory for cache results
+  auto rans_symbols_tensor =
+      CREATE_TENSOR(num_symbols * sizeof(RansSymbol), torch::ScalarType::Byte,
+                    symbols.device());  // store encoded symbols;
+  RansSymbol *rans_symbols = (RansSymbol *)rans_symbols_tensor.data_ptr();
+  auto results_tensor = CREATE_TENSOR(
+      num_symbols * sizeof(RansCUDAKernelResult), torch::ScalarType::Byte,
+      symbols.device());  // store whether value == max_value;
+  RansCUDAKernelResult *results =
+      (RansCUDAKernelResult *)results_tensor.data_ptr();
+  // get encoded symbols
+  rans_encode_with_indexes_cuda_kernel<<<blocks, threads, 0, stream>>>(
+      symbols.data_ptr<int>(), indexes.data_ptr<int>(), cdfs.data_ptr<int>(),
+      cdfs_sizes.data_ptr<int>(), offsets.data_ptr<int>(), num_symbols,
+      num_cdfs, max_cdf_bin_size, rans_symbols, results);
+  // sync
+  cudaStreamSynchronize(stream);
```

Comment:
cudaStreamSynchronize is not necessary since tensor.cpu would automatically sync the stream.